content/post/kmcuda4.md

```diff
@@ -0,0 +1,489 @@
+---
+author: vadim
+date: 2016-11-21
+title: "kmcuda (K-Means on GPU) version 4 is released"
+draft: false
+image: /post/kmcuda4/nvprof.png
+description: "Our kmcuda v4 is released, featuring multi-gpu, float16, Spherical K-Means and improved precision."
+---
+
+Some time ago, I wrote the article about [src-d/kmcuda](https://github.com/src-d/kmcuda)
+named [Towards Yinyang K-means on GPU](http://blog.sourced.tech/post/towards_kmeans_on_gpu/).
+This weekend, we released version 4 of that library. Here is the brief list of
+changes:
+
+1. K-Means iterations can run in parallel on several GPU cards, so
+the elapsed time reduces proportional to the number of GPUs.
+2. Type of the input samples can be [16-bit floating point](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) (fp16),
+so the maximum size of clustered data doubles.
+3. Distance between points can be calculated with the [angular metric](https://en.wikipedia.org/wiki/Cosine_similarity#Angular_distance_and_similarity),
+thus the clustering algorithm becomes a variant of [Spherical K-Means](http://www.google.com/search?q=spherical+k-means).
+4. The precision of the calculation of distances was improved at the expense
+of some small computational overhead thanks to [Kahan summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm).
+5. Zero-copy device input and output, that is, taking samples from GPU
+memory and writing the results to the preallocated GPU memory.
+6. Build was much improved: selection of the CUDA device architecture, optional components,
+continuous integration, [PyPi manylinux wheels](https://github.com/pypa/manylinux).
+7. Extensive test suite was developed.
+
+I will go into details about each of the points.
+
+### Multi-GPU
+
+![nvprof](/post/kmcuda4/nvprof.png)
+
+Clustering on several GPUs at the same time is not straightforward,
+because you cannot simply split the samples into equal parts and
+feed them in parallel. The calculation of the centroifs requires
+all the data in the same place. While there can be workarounds for
+storing the whole dataset in the memory, I chose a quicker to implement
+and more performant solution.
+
+Let me remind how Lloyd algorithm works:
+
+1. We calculate the distances between the samples and the centroids,
+pick the centroid with the minimum distance fo every sample (assignment stage).
+2. We update the centroids by averaging all the samples assigned to each
+(centroids stage).
+
+The current multi-gpu scheme is as follows.
+
+1. We cut the samples into the same number of intervals as the number of
+GPUs. Each GPU works with one interval, calculates the distances,
```

Comment:
consider rewording the first sentence in the paragraph to:

"We cut the samples into as many intervals as GPUs are available."